(PUP-8080) Add tests for i18n, puppet, and modules

ericldelaney opened the pull request:
This is an initial set of tests for testing i18n and modules
The puppet_agent_with_module.rb test fails based on the current agent

ericldelaney commented:
@Magisus here are the tests I've got so far the puppet apply tests I would expect to pass, the puppet agent test I think should work, correctly, but the puppet fact test is a work in progress so far.

Magisus commented:
Thanks, I'll try these out as I continue working through my module translation work.

ericldelaney commented:
@Magisus I've updated the tests and added a cached catalog version.

@eputnam @MikaelSmith Here are some new tests, the puppet apply tests work, but the puppet agent agent puppet facts tests I failing but are my best guess. What other tests would you like to see?

MikaelSmith commented on acceptance/tests/i18n/modules/puppet_agent_with_module.rb:
```diff
@@ -0,0 +1,145 @@
+        # TODO
+        # TODO
+        # TODO are we supposed to get a Server 500 on SERVER: error here?
+        # Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Function Call,
```
Not sure, but it should be translated (both "Server Error" and the "Failed to create").

eputnam replied:
Yes, this failure is on purpose. They should both be translated, the "failed to create" message comes from the module in init.pp.

puppetcla commented:
CLA signed by all contributors.

MikaelSmith commented:
I'll spend time on this Monday.

eputnam commented on acceptance/lib/puppet/acceptance/i18ndemo_utils.rb:
```diff
@@ -0,0 +1,35 @@
+
+    I18NDEMO_MODULE_NAME = "eputnam-i18ndemo"
+
+    # TODO switch to puppetlabs/beaker-i18n_helper
```
also, if you've got better ways to do this stuff or more ideas for the helper, _please_ don't hesitate to put up a PR or 10

MikaelSmith commented:
One case missing from this is a module with translations that are not english or japanese, in the locale of that language. This is to test that we correctly get module translations in that language even if Puppet doesn't have translations, and that Puppet falls back to english.

MikaelSmith commented:
I'd also like to see a `puppet resource` test, that would be somewhat similar to `puppet apply`, triggering an error message from the type or provider.

MikaelSmith commented:
I'd also add an application or face to the module with translations, and test that it's appropriately translated.

ericldelaney commented:
@eputnam Are you planning to make a module that is not english or Japanese?

MikaelSmith commented:
Actually, it could probably be the same module. But then you want to test in that locale so that Puppet defaults to english but the module uses it's translation.

ericldelaney commented:
@Magisus @MikaelSmith I've updated the tests to skip the cached catalog on Windows (PUP-8368) and I still need to uncomment and update some test code once we have translations. But this is what I'd like to get merged and running in CI now.

MikaelSmith commented:
Ok. Are the other test scenarios I mentioned going to be documented somewhere for future work?

ericldelaney commented:
@MikaelSmith Did I miss something? 
You wanted a different language (not English or Japanese), that's here:
https://github.com/puppetlabs/puppet/pull/6349/files#diff-dfa8b1237494ddb51d3f658576f9b96fR18

You wanted puppet resource with an provider error, that's here:
https://github.com/puppetlabs/puppet/pull/6349/files#diff-a2fde07aecfd5980cdd526f01ef055b5R36

Is there something else I'm missing?

MikaelSmith commented:
Ah, you didn't mention those in your comment so I wasn't sure you'd done anything with them. I'll take a closer look next week.

MikaelSmith commented:
jenkins please test this

ericldelaney commented:
@Magisus Could I get a review and merge of this PR. All of the tests pass on all platforms (tested that on Friday). The 2 outstanding issues are separate tickets to be addressed, but I wanted to get these passing tests checked in and giving us coverage.

Magisus commented:
Do we want to wait for that last run to finish?
